File: fs/before_energy/154__find-minimum-in-rotated-sorted-array-ii/cpp/find-minimum-in-rotated-sorted-array-ii_benchmark_v2.py
# Time:  O(logn) ~ O(n)
# Space: O(1)

class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        left, right = 0, len(nums) - 1
        while left < right:
            mid = left + (right - left) / 2

            if nums[mid] == nums[right]:
                right -= 1
            elif nums[mid] < nums[right]:
                right = mid
            else:
                left = mid + 1

        return nums[left]


class Solution2(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        left, right = 0, len(nums) - 1
        while left < right and nums[left] >= nums[right]:
            mid = left + (right - left) / 2

            if nums[mid] == nums[left]:
                left += 1
            elif nums[mid] < nums[left]:
                right = mid
            else:
                left = mid + 1

        return nums[left]


if __name__ == "__main__":
    tests = [
        [1],
        [2, 2, 2, 2, 2],
        [3, 4, 5, 1, 2],
        [2, 2, 2, 3, 1, 2],
        [10, 1, 10, 10, 10],
        [1, 2, 3, 4, 5, 6, 7],
        [2, 2, 3, 4, 4, 5, 1, 1, 2],
        [5, 6, 7, 0, 1, 2, 3, 4],
        [4, 4, 4, 5, 6, 7, 0, 1, 2, 2, 3],
        [0, 0, 1, 2, 2, 2, 2, 2, 0]
    ]

    sink = 0
    sol = Solution()

    iterations = 1000
    for iter in xrange(iterations):
        #checksum = 0
        for t in tests:
            r = sol.findMin(t)
        #sink = checksum

File: fs/cpp_before_checksum/154__find-minimum-in-rotated-sorted-array-ii/find-minimum-in-rotated-sorted-array-ii_benchmark_v2.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int findMin(std::vector<int>& nums) {
        int left = 0, right = static_cast<int>(nums.size()) - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == nums[right]) {
                --right;
            } else if (nums[mid] < nums[right]) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return nums[left];
    }
};

int main() {
    std::vector<std::vector<int>> tests = {
        {1},
        {2, 2, 2, 2, 2},
        {3, 4, 5, 1, 2},
        {2, 2, 2, 3, 1, 2},
        {10, 1, 10, 10, 10},
        {1, 2, 3, 4, 5, 6, 7},
        {2, 2, 3, 4, 4, 5, 1, 1, 2},
        {5, 6, 7, 0, 1, 2, 3, 4},
        {4, 4, 4, 5, 6, 7, 0, 1, 2, 2, 3},
        {0, 0, 1, 2, 2, 2, 2, 2, 0}
    };

    volatile int sink = 0;
    Solution sol;

    const int iterations = 1;
    for (int iter = 0; iter < iterations; ++iter) {
        int checksum = 0;
        for (auto& t : tests) {
            checksum += sol.findMin(t);
        }
        sink += checksum;
        

    } std::cout << sink << std::endl;

    return 0;
}